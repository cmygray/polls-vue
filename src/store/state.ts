import { Poll } from '@/models';

export interface State {
  polls: {
    [id: string]: Poll;
  };
}

export const state: State = {
  polls: {},
};
