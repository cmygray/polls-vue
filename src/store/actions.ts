import { ActionTree } from 'vuex';

import { State } from '@/store/state';
import { Poll } from '@/models';
import { createPoll, fetchPolls } from '@/apis';

export const actions: ActionTree<State, State> = {
  async fetchPolls({ commit }): Promise<Poll[]> {
    const { data } = await fetchPolls();

    data.forEach(poll => {
      commit('SET_POLL', poll);
    });

    return data;
  },

  async createPoll({ commit }, title): Promise<Poll> {
    const { data } = await createPoll(title);

    commit('SET_POLL', data);

    return data;
  },
};
