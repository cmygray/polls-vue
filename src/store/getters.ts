import { GetterTree } from 'vuex';
import { State } from '@/store/state';

export const getters: GetterTree<State, State> = {
  allPolls: state =>
    Object.values(state.polls).sort(
      (currentPoll, nextPoll) =>
        +new Date(nextPoll.created_at) - +new Date(currentPoll.created_at)
    ),
  getChoicesOfPoll: state => (pollId: string) =>
    Object.values(state.choices)
      .filter(choice => choice.poll_id === pollId)
      .sort((current, next) => +next.id - +current.id),
};
