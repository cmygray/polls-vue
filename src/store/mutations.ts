import Vue from 'vue';
import { MutationTree } from 'vuex';

import { Poll } from '@/models';
import { State } from '@/store/state';

export const mutations: MutationTree<State> = {
  SET_POLL(state, poll: Poll) {
    Vue.set(state.polls, poll.id, poll);
  },
};
