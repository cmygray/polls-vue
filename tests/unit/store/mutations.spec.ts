import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { poll } from '../__fixtures__';

describe('mutations', () => {
  describe('SET_POLL', () => {
    it('should set poll to expected path', () => {
      const PATH = poll.id

      mutations.SET_POLL(state, poll);

      expect(state).toMatchSnapshot();
      expect(state.polls[PATH]).toEqual(poll)
    });
  });
});
