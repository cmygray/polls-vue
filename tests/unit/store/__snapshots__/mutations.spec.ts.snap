// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations SET_POLL should set poll to expected path 1`] = `
Object {
  "polls": Object {
    "28": Object {
      "choices": Array [
        Object {
          "choice_text": "항목을 입력하세요",
          "id": "5",
          "poll_id": 28,
          "votes_count": 0,
        },
        Object {
          "choice_text": "항목을 입력하세요",
          "id": "6",
          "poll_id": 28,
          "votes_count": 0,
        },
        Object {
          "choice_text": "항목을 입력하세요",
          "id": "7",
          "poll_id": 28,
          "votes_count": 0,
        },
      ],
      "created_at": "2019-06-18T15:28:29.909356",
      "ends_at": "2019-06-21T00:00:00",
      "id": "28",
      "poll_title": "TEST",
      "starts_at": "2019-06-18T15:28:29.909377",
    },
  },
}
`;
